lib/timers.js

```diff
@@ -796,42 +768,39 @@ function tryOnImmediate(immediate, oldTail) {
 
 function runCallback(timer) {
   const argv = timer._argv;
-  const argc = argv ? argv.length : 0;
   if (typeof timer._onImmediate !== 'function')
     return promiseResolve(timer._onImmediate, argv[0]);
-  switch (argc) {
-    // fast-path callbacks with 0-3 arguments
-    case 0:
-      return timer._onImmediate();
-    case 1:
-      return timer._onImmediate(argv[0]);
-    case 2:
-      return timer._onImmediate(argv[0], argv[1]);
-    case 3:
-      return timer._onImmediate(argv[0], argv[1], argv[2]);
-    // more than 3 arguments run slower with .apply
-    default:
-      return Function.prototype.apply.call(timer._onImmediate, timer, argv);
-  }
+  if (!argv)
+    return timer._onImmediate();
+  functionApply.call(timer._onImmediate, timer, argv);
 }
 
 
-function Immediate() {
-  // assigning the callback here can cause optimize/deoptimize thrashing
-  // so have caller annotate the object (node v6.0.0, v8 5.0.71.35)
-  this._idleNext = null;
-  this._idlePrev = null;
-  this._onImmediate = null;
-  this._argv = null;
-  this._onImmediate = null;
-  this._destroyed = false;
-  this.domain = process.domain;
-  this[async_id_symbol] = ++async_id_fields[kAsyncIdCounter];
-  this[trigger_async_id_symbol] = initTriggerId();
-  if (async_hook_fields[kInit] > 0)
-    emitInit(
-      this[async_id_symbol], 'Immediate', this[trigger_async_id_symbol], this
-    );
+class Immediate {
+  constructor(args) {
```

Comment:
What if we allow the `_onImmediate` value to be passed to the constructor to save a step?

Reply:
See the comment right below... that's what I wanted to do but it runs 3x slower on our benchmarks. I have not figured out what exactly causes it to plummet.

Reply:
Ah ok, if I had to guess I would say it might be the different types (`Promise` vs function). \*shrug\*

Reply:
But then it's fine for Timeout (and it was in its constructor originally)... so I have no clue. Maybe @bmeurer can shed some light on this for us?

Reply:
My guess here is that this trigger function tracking on the hidden class. Have you tried passing it to constructor, but setting the field `_onImmediate` to `null` before and only then store the value?

Reply:
Will try, thanks for the tip!

Reply:
That did it!

Reply:
Hey 👋 
I know that it's a pretty old PR, but I was studying this code, and these lines caught my attention:
```js
// this must be set to null first to avoid function tracking
// on the hidden class, revisit in V8 versions after 6.2
this._onTimeout = null;
this._onTimeout = callback;
```

Why in these cases the hidden class optimization is spoiling the performance instead of improving?
Is it still happen?